### ISSUE TITLE
Fixed module set -> Chain of responsibility migration

### ISSUE THREAD

**Issue by hashedone** [CosmWasm/cw-multi-test#7]
People are facing the problem that if they have a contract emitting IBC messages, it is technically impossible to test them - passing any IBC message through a router will always cause panic. It is semi-solved for staking/custom messages by a separate module for them to fail by default, but possible to overwrite their behaviour. For now, the problem exists for IBC, Stargate and Gov messages (so basically anything introduced via a `stargate` feature).

Another problem is how staking is implemented - it just enables the `stargate` feature flag by default (enabling relevant features in `cosmwasm-std`, but what would happen if someone would explicitly disable it - via `no-default-fetures`? It would fail to compile because of a lack of staking-related types in cw-std).

This approach of the fixed module set has proven to be difficult to maintain properly. I have another idea. Let's model a module like something accepting any cosmos message and handling it optionally - if it knows how to handle it, it will return a proper result. If not - it would return some `None` equivalent. If any module manages to handle the message, its processing is finished successfully.

To make it easy, I would create two generic "modules" or better "handlers". `DefaultHandler` would take any message and report an "unhandled message" error. One called on the very end of a chain. The other one would be `FallbackHandler<Module, Fallback>`. The `FallbackHandler` tries to handle a message using `Module`, and if it fails, it delegates handling it to `Fallback`. `Fallback` is typically another `FallbackHandler` or `DefaultHandler` at the end - basically a simple linked list.

Now to make it even better API-wise, `FallbackHandler` would take an even better form: `FallbackHandler<T, Filter<T>, Module<T>, Fallback>`, where `Filter` takes a `Cosmos` message, returning some `Option<T>`, and `Module<T>` is a handler taking a `T` (eg. `BankMsg`) handling it. If the filter returns `Some`, then only my `Module` is allowed to handle the message. Otherwise, it is forwarded further (so it is shortcut-break if there is an unsupported bank message in future).

This also allows changing the handler for a particular message - as they are unique identified with `T` - it would allow keeping the same Api we have now with some defaults (bank, wasmkeeper), which can be substituted with custom messages. It would also allow easily feature-enable some other implementations in future or allow others to provide custom implementation of modules not yet implemented in MT at all, under some feature flags.

Any thoughts? I am happy to provide PoC of that.

**Comment by gmanev7** [comment]
@hashedone, we have recently stuck at that limitation of cw-milti-test. Our contract sends an IBC transfer message that causes panic. Thanks for filing this ticket!
The proposed solution looks more generic than the current one with the router to a fixed set of handlers/modules. That solution would allow us to plug an implementation of a handler of IBC messages, either some provided with the cw-multi-test, or our own,

Two points here: 
- Filter<T> could be made part of the interface of Module<T> to express better their close relationships ... otherwise one may provide a filter that does not behave consistently with the module. After all, they share the same responsibility - if and how to handle a message
- did not get how a unique handler for T could be guaranteed?

**Comment by hashedone** [comment]
The idea of splitting the module and filter is very much by design. I want that if someone implements a "bank module", his module has no idea of the `CosmosMsg` or anything besides `BankMsg`. Bank module POV should end on the `BankMsg`. The idea of the `Filter` is to cut this `CosmosMsg` part relevant to a particular target.

In practice, I think that `T` would be coupled with `Filter` with something like `MsgDescirptor`. For example:

```rust
trait MsgDescriptor<CustomMsg, CustomQuery> {
  type Msg;
  fn filter(&self, msg: &CosmosMsg<CustomMsg, CustomQuery>) -> Option<Msg>;
}
```

Then example `BankDescriptor` would be:

```rust
struct BankDescriptor;

impl<M, Q> MsgDescriptor<M, Q> for BankDescriptor {
  type Msg = BankMsg;
  fn filter(&self, msg: ...) -> ... { ... }
}
```

It could be improved even more (by ownership transfer of msg), but it doesn't matter right now (implementation detail).

The module trait would be like:

```rust
trait Module<Msg, Query, Sudo> {
  fn execute(msg: Msg) -> Response;
  fn query(msg: Query) -> Response;
  fn sudo(msg: Sudo) -> Response;
}
```

so basically the same as it is (I know types are bit different - this is just visualisation). Then `BankKeeper` is just implementing this `Module<BankMsg, BankQuery, Empty>`.

The uniqueness is guaranteed by construction. Basically, the whole `DefaultHanlder` and `FallbackHanlder` are implementation details (and their common `Handler` trait is a sealed trait), and those are hidden from the user. Adding a handler would "iterate" through the whole list to ensure that a particular type is not yet added (in CT). Also, for "updating" a handler, it would make sure it is actually added and alters it. Everything is possible to be done in compile time.

**Comment by ethanfrey** [comment]
This is a lot to digest. I can do into details on the implementation, but let me first go over the goal. We may also want to revisit all the generics we have everywhere before adding more complexity. I like the design in general, just wondering about the extra complexity on a module that only two people really understand fully (you and I).

Anyway, let's look at the issues:

1. We expose a way to decompose CosmosMsg into some variants and pass them to different "Modules". We support a subset, and we also may have to make some optional based on the feature flags enabled on cosmwasm-std.
2. If you want to add another module, you need to update a number of places in the code, and cannot do it without forking the module. If we add support for all existing variants of CosmosMsg, not sure if there will be the need externally, but it is good the architecture doesn't block 3rd party extensions
3. In particular, people want full support for IBC tests (both sides), which is a known issue https://github.com/CosmWasm/cw-multi-test/issues/1 This would be a good first step to making that possible (ideally possible to add later without forking cw-multi-test)
4. Note that mocking StargateMsg/Query is never needed in practice. People are working frameworks to test in full nodes as you need to test that against the real Go implementation there.

**Comment by ethanfrey** [comment]
The TL;DR of the design would be that instead of doing a match on `execute`, like we do here: https://github.com/CosmWasm/cw-multi-test/blob/daeb37dc4cc15efea599b5219e11d47d607c82a8/src/app.rs#L788-L806 we would iterate over `Vec<Box<dyn MsgDescriptor >>` or something like that and execute them on matches.

It does seem more extensible in a generic sense and I like the idea for a fully extensible message format. But it seems overkill if we really only need to add one more type here (okay, maybe 2 if anyone wants to test gov). I would keep this idea around for the future, but not implement now in the context of IBC.

Also, adding IBC modules is not just about execute and query, but allowing calling into those other 6 entry points in a contract and will need larger changes. This refactor wouldn't really help this.

**Comment by hashedone** [comment]
> Also, adding IBC modules is not just about execute and query, but allowing calling into those other 6 entry points in a contract and will need larger changes.
That is not fully true. It is needed to properly test contracts' IBC functionality. But it is not the most important case. The important case is: "I want to test my contract which triggers IBC, but I want to test only non-IBC part - I want to ignore any IBC messages send by contract".

The problem is that now if you trigger a message which performs some on-chain things and additionally sends tokens via IBC, you cannot technically test this on-chain functionality on MT at all - because `WasmKeeper` would try to execute all messages from the response, including IBC, and Router would fail on it.

Technically there is an additional problem already introduced - stargate is improperly feature-gated in `Cargo.toml` - If you disable the `staking` feature (which you can - it is under a feature flag), you fail to compile because you also disable a feature on cosmwasm-std, and the big match would be invalid (as there would be no `Staking` variant on message). It never occurred, as the feature is enabled by default, but it is not a nice Api.

Basically, the problem is - if we add a feature flag - for now, it is staking, but now it would be ibc (not stargate as it would be only part of stargate, I think). Then you do it to allow users to pick used flags. And compilation should never fail because of disabling flags.

However, in the current design, it would be extremely difficult to solve this feature gating problem correctly. The reason is that you need to remove generics for particular modules everywhere, depending only on feature flags. And finally, what set generics are to be there would depend on enabled feature flags. This API looks terrible, and documentation would also be very bad for this. So we either have to find a better design (proposed by me would work, but it has other downsides I figured out this evening), or we should remove feature-gating MT (std still can be feature-gated, but MT framework would always assume the existence of all modules - just failing ones if they are not "present").

About downsides - the one I don't know how to deal with at all is backward compatibility. The big thing is module initialization - on the build, there is a function that takes router, state and API objects. This is not perfect, but working, as if I want to initialize the bank module, I reach to bank keeper in the router and perform actions on it. It would be impossible with my design as modules would not be easy to random-access this way. I have a couple of ideas on how to solve it, but it would require completely breaking this API without the possibility of an easy transition. All solutions I have for that are so much different that they cannot have any compatibility layer with the previous approach. I don't like this fact at all. I need to think about it more.

Probably I would first try to add an Ibc module like the Staking one, but this is also not this trivial (nor is it difficult - just much tedious work). We can talk about this Wed.

**Comment by ethanfrey** [comment]
> or we should remove feature-gating MT (std still can be feature-gated, but MT framework would always assume the existence of all modules - just failing ones if they are not "present").

That should be fine, as it will only affect test code, and the production code should ever match over all CosmosMsg (right?). They just construct them and those will still work when we add more variants.

**Comment by ethanfrey** [comment]
I thought of another reason to not do your approach. Many times to set up tests, we want to access `&mut router.bank` or something like that with the underlying BankKeeper type (not the dynamic trait). If we make a list like your plan, we can never cleanly recover the underlying type.
